## 4.5 Data Consistency Checks

### This script contains the following points:

#### 1. Create data to experiment on
#### 2. Check for null values
#### 3. Check for duplicates
#### 4. Check for mixed-type data
#### 5. Create flag for no previous order

In [1]:

##import libraries
import pandas as pd
import numpy as np
import os

In [17]:
##  import the “products.csv” data set from your “Original Data” folder as df_prods and your “orders.csv” data set from your “Prepared Data” folder as df_ords
path = r'/Users/rachelallen/CareerFoundry Achievement 4/08-2022 Instacart Basket Analysis/02 Data'
df_prods = pd.read_csv(os.path.join(path, 'Original Data', '4.3_orders_products', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
## create a dataframe to prep and practice
df_test = pd.DataFrame()

In [5]:
##create a mixed type column
df_test['mix'] = ['a','b', 1, True]

In [6]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [7]:
##check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
##change to string
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
##change to int

In [18]:
##check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
##create subset with null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
df_prods.shape

(49693, 5)

In [23]:
##create new df without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
##check to make sure it worked
df_prods_clean.shape

(49677, 5)

In [27]:
##create new df of only duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [28]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
##check how many rows before dropping duplicates
df_prods_clean.shape

(49677, 5)

In [32]:
##create new clean dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
df_prods_clean_no_dups.shape

(49672, 5)

In [34]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_prepared.csv'))

In [35]:
# Task 4.5
## 2: Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

the max price is 99999? seems very strange, let me check on that really quick

In [38]:
df_prods[df_prods['prices'] == 99999.0]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

I would send this to the client and ask if this is correct 

In [39]:
## Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)


In [40]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour                  int64
days_since_prior_order    float64
dtype: object

4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

My code didn't show a column with a mixed data type

In [41]:
##5: Run a check for missing values in your df_ords dataframe

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour                     0
days_since_prior_order    206209
dtype: int64

Looks like days_since_prior_order is full of blanks. I'd propose that that might be first orders -- if there is no prior order they might be labeled "nan"

In [42]:
df_ords.shape

(3421083, 7)

In [45]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week    order_hour  days_since_prior_order  
count       3.421083e+06  3.421083e+06            3.214874e+06  
mean        2.776219e+00  1.345202e+01            1.111484e+01  
std         2.046829e+00  4.226088e+00            9.206737e+00  
min         0.000000e+00  0.000000e+00            0.000000e+00  
25%         1.000000e+00  1.000000e+01            4.000000e+00  
50%         3.000000e+00  1.300000e+01            7.000000e+00  
75%         5.000000e+00  1.600000e+01            1.500000e+01  
max         6.000000e+00  2.300000e+01            3.000000e+01

In [46]:
df_ords[df_ords['days_since_prior_order'] == 0]

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
8                 8   2295261        1             9                  1   
43               43   2557754        4             5                  5   
189             189   2148135       17            30                  6   
296             296    965160       24            19                  0   
337             337   3162630       27            24                  2   
...             ...       ...      ...           ...                ...   
3420980     3420980      4277   206206            47                  2   
3420985     3420985    648225   206206            52                  0   
3420991     3420991   2475308   206206            58                  0   
3421001     3421001   1043943   206206            68                  0   
3421044     3421044   2079142   206208            26                  2   

         order_hour  days_since_prior_order  
8                16                     0.0  
43               13                     0.0  
189              15                     0.0  
296              16                     0.0  
337              15                     0.0  
...             ...                     ...  
3420980          17                     0.0  
3420985          19                     0.0  
3420991          15                     0.0  
3421001          20                     0.0  
3421044          16                     0.0  

[67755 rows x 7 columns]

I had been thinking that "nan" in days_since_prior_order might mean no previous order. There is also a minimum of zero -- so either zero means "placed another order today" OR zero might mean no previous order.

In [70]:
## 6 Address the missing values 
## with some help from Stack Overflow and google I have created a new column to flag NaN as "no previous order"
df_ords['flagged'] = np.where(df_ords['days_since_prior_order'].notna(), 'No Previous Order', 'Previous Order')

In [69]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  order_hour  \
0           0   2539329        1             1                  2           8   
1           1   2398795        1             2                  3           7   
2           2    473747        1             3                  3          12   
3           3   2254736        1             4                  4           7   
4           4    431534        1             5                  4          15   

   days_since_prior_order            flagged  
0                     NaN     Previous Order  
1                    15.0  No Previous Order  
2                    21.0  No Previous Order  
3                    29.0  No Previous Order  
4                    28.0  No Previous Order

Rather than use the mean or the median, I decided to create a flag column so it wouldn't change any of the data. Now I can filter for only orders with previous orders, or not. 

In [71]:
## count number of previous orders to make sure it's correct
df_ords['flagged'].value_counts(dropna = False)

No Previous Order    3214874
Previous Order        206209
Name: flagged, dtype: int64

In [72]:
## that check told me I did it backwards -- 206209 is the number of NO previous orders. Whoops! Changing that code really fast.


df_ords['flagged'] = np.where(df_ords['days_since_prior_order'].notna(), 'Previous Order', 'No Previous Order')

In [73]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  order_hour  \
0           0   2539329        1             1                  2           8   
1           1   2398795        1             2                  3           7   
2           2    473747        1             3                  3          12   
3           3   2254736        1             4                  4           7   
4           4    431534        1             5                  4          15   

   days_since_prior_order            flagged  
0                     NaN  No Previous Order  
1                    15.0     Previous Order  
2                    21.0     Previous Order  
3                    29.0     Previous Order  
4                    28.0     Previous Order

In [74]:
## 7: Run a check for duplicate values in your df_ords data
df_ords_dups = df_ords[df_ords.duplicated()]

In [76]:
df_ords_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour, days_since_prior_order, flagged]
Index: []

## 8: It doesn't look like there are any duplicates in df_ords but I'm going to drop duplicates just in case

In [77]:
df_ords.shape

(3421083, 8)

In [78]:
df_ords_clean = df_ords.drop_duplicates()

In [79]:
df_ords_clean.shape

(3421083, 8)

In [80]:
## 9: explort cleaned df_ords
df_ords_clean.to_csv(os.path.join(path, 'Prepared Data', 'orders_prepared.csv'))